### PR TITLE
[reland] Add sharding tests to multigpu-test.sh and fix custom operator decorator

### PR DESCRIPTION
**Pull request by pritamdamania87**
1. Enabled multigpu tests.
2. Fixed failing multigpu tests.
3. Fixed custom operator decorator to be first preference in operator dispatch.

### CODE REVIEW

**Review by fduwjj — APPROVED** [review]

**Review by wanchaol — APPROVED** [review]

File: .jenkins/pytorch/multigpu-test.sh [hunk 1]
```diff
@@ -28,4 +28,27 @@ time python test/run_test.py --verbose -i distributed/test_c10d_spawn_nccl
 time python test/run_test.py --verbose -i distributed/test_store
 time python test/run_test.py --verbose -i distributed/test_pg_wrapper
 time python test/run_test.py --verbose -i distributed/rpc/cuda/test_tensorpipe_agent
+time python test/run_test.py --verbose -i distributed/_shard/checkpoint/test_checkpoint
+time python test/run_test.py --verbose -i distributed/_shard/checkpoint/test_file_system_checkpoint
+time python test/run_test.py --verbose -i distributed/_shard/sharding_spec/test_sharding_spec
+time python test/run_test.py --verbose -i distributed/_shard/sharding_plan/test_sharding_plan
+time python test/run_test.py --verbose -i distributed/_shard/sharding_spec/test_sharding_spec
+time python test/run_test.py --verbose -i distributed/_shard/sharding_plan/test_sharding_plan
```
Comment (janeyx99) [inline .jenkins/pytorch/multigpu-test.sh:36]:
```suggestion
```
repeated

File: .jenkins/pytorch/multigpu-test.sh [hunk 2]
```diff
@@ -28,4 +28,27 @@ time python test/run_test.py --verbose -i distributed/test_c10d_spawn_nccl
...
+time python test/run_test.py --verbose -i distributed/_shard/sharded_tensor/ops/test_elementwise_ops
+time python test/run_test.py --verbose -i distributed/_shard/sharded_tensor/ops/test_embedding
+time python test/run_test.py --verbose -i distributed/_shard/sharded_tensor/ops/test_embedding_bag
+time python test/run_test.py --verbose -i distributed/_shard/sharded_tensor/ops/test_binary_cmp
+time python test/run_test.py --verbose -i distributed/_shard/sharded_tensor/ops/test_init
+time python test/run_test.py --verbose -i distributed/_shard/sharded_tensor/ops/test_linear
+time python test/run_test.py --verbose -i distributed/_shard/sharded_tensor/ops/test_math_ops
+time python test/run_test.py --verbose -i distributed/_shard/sharded_tensor/ops/test_matrix_ops
+time python test/run_test.py --verbose -i distributed/_shard/sharded_tensor/ops/test_softmax
+time python test/run_test.py --verbose -i distributed/_shard/sharding_spec/test_sharding_spec
```
Comment (janeyx99) [inline .jenkins/pytorch/multigpu-test.sh:50]:
```suggestion
```
also repeated

**Review by janeyx99 — CHANGES_REQUESTED** [review]
I still see some duplication

**Comment by pritamdamania87** [comment]
@janeyx99 Fixed them in the latest update

**Review by janeyx99 — APPROVED** [review]
Looks good code wise! 

Please wait for the multigpu tests to pass before landing :D

**Comment by pritamdamania87** [comment]
@janeyx99 Lint seems to be failing due to an unrelated reason:

```
>>> Lint for .github/scripts/trymerge.py:

  Error (MYPY) [comparison-overlap]
    Non-overlapping equality check (left operand type: "Tuple[str, str]",
[2022-05-20T22:31:02Z DEBUG lintrunner::persistent_data] Writing run info to /home/runner/.local/share/lintrunner/a37df4630fe1f5acff882ee64ccae4fb5dc8d95c582db33d3b69846b93567c2c/runs/2022-05-20T22:29:05.011Z_0483e62621e5043a5d98eefbd9eb26338128bcae37b14d65aadd9b075982a9ab
    right operand type: "Literal['SUCCESS']")

        907  |                for check in checks:
        908  |                    if checks[check] is None:
        909  |                        pending_checks.append(check)
    >>> 910  |                    elif checks[check] != 'SUCCESS':
        911  |                        failed_checks.append(check)
        912  |                if len(failed_checks) > 0:
        913  |                    raise RuntimeError(f"Failed to merge due to failing checks, {','.join(failed_checks)}")
```

Is it ok to merge this using "@pytorchbot merge -f"?

**Comment by janeyx99** [comment]
@pritamdamania87 looks like a rebase fixed the lint issue + you've landed successfully, let me know if there's more I should look into.

**Pull request closed without merging**